File: java/src/main/java/org/revenj/security/PermissionManager.java
package org.revenj.security;

import org.revenj.patterns.DataSource;
import org.revenj.patterns.Query;
import org.revenj.patterns.Specification;

import java.io.Closeable;
import java.lang.reflect.Array;
import java.security.Principal;
import java.util.Arrays;
import java.util.Collection;

public interface PermissionManager {
	boolean canAccess(String identifier, Principal user);

	default boolean canAccess(Class<?> manifest, Principal user) {
		return canAccess(manifest.getTypeName(), user);
	}

	<T extends DataSource> Query<T> applyFilters(Class<T> manifest, Principal user, Query<T> data);

	<T extends DataSource> Collection<T> applyFilters(Class<T> manifest, Principal user, Collection<T> data);

	<T extends DataSource> Closeable registerFilter(Class<T> manifest, Specification<T> filter, String role, boolean inverse);
}

File: scala/revenj-core/src/main/scala/net/revenj/security/PermissionManager.scala
package net.revenj.security

import java.security.Principal

import scala.reflect.runtime.universe._

trait PermissionManager {
  def canAccess(identifier: String)(implicit user: Principal): Boolean

  def canAccess[T : TypeTag](implicit user: Principal): Boolean = canAccess(typeOf[T].toString)

  def invalidate(): Unit
}
